### PR TITLE
target/ppc: implement pdepd/vpdepd and pextd/vpextd

### PR DESCRIPTION
**Pull request by mferst**
Signed-off-by: Matheus Ferst <matheus.ferst@eldorado.org.br>

### CODE REVIEW

File: target/ppc/translate/vector-impl.c.inc
Comment (lffpires) [inline target/ppc/translate/vector-impl.c.inc:36]: Since this is now a generic utility function for dealing with Vector Facility instructions and VX-form, maybe rename 'src' and 'mask' to generic names (arg1+arg2, arg_a+arg_b, or something similar)

Comment (lffpires) [reply target/ppc/translate/vector-impl.c.inc:36]: I would also put the change to VCFUGED and the introduction of do_call_vx_doubleword_helper in a separate commit, unrelated to vpdepd/vpextd.

File: target/ppc/translate/vector-impl.c.inc
Comment (lffpires) [inline target/ppc/translate/vector-impl.c.inc:58]: Should we REQUIRE_64BIT and wrap the call to do_call_vx_doubleword_helper with #if defined(TARGET_PPC64)/qemu_build_not_reached() too?

I've been meaning to ask you this before, but ended up forgetting to.

Without these 2, aren't we being less restrictive with vcfuged than with cfuged? (this applies to the other vector x non-vector instructions that share helpers too)

**Review by lffpires — APPROVED** [review]
The changes look good. Just made a few minor comments.

Regarding the split between commits (both on this PR and my own), I'm not against keeping instructions that are very closely related in the same commit (pdepd/pextd, cntlzdm/cnttzdm, etc.).
Personally, I even find it slightly easier to review them at the same time. Do you feel strongly about always keeping them separate or are you okay either way?
This already happened for other instruction groups (addi/addis/paddi, cmp/cmpi/cmpl/cmpli, etc.)

File: target/ppc/int_helper.c
Comment (mferst) [reply target/ppc/int_helper.c:387]: Ah, it's a left over from my initial implementation, I'll change that.

File: target/ppc/translate/vector-impl.c.inc
Comment (mferst) [reply target/ppc/translate/vector-impl.c.inc:36]: At first glance it seemed that all insn that could use this method had source and mask registers, but a more generic name might be future proof, Ill change that.

I did this in a single commit to avoid introducing an "useless" method, but I can I split it too.

File: target/ppc/translate/vector-impl.c.inc
Comment (mferst) [reply target/ppc/translate/vector-impl.c.inc:58]: The ifdef is not required here. It's used in cfuged/pdepd/pextd to avoid a compilation error for ppc-softmmu and ppc-linux-user targets, where `cpu_gpr` is `TCGv` (i.e. `TCGv_i32` for those targets). Even though the code after REQUIRE_64BIT is optimized, the compiler checks the types, and `gen_helper` receives `TCGv_i64`.

(We could avoid the error by declaring the helper receiving `target_ulong`, but IIRC vector registers are 128-bits even in 32-bits, so it would lead to wrong results.)

About REQUIRE_64BIT, I'm not sure. The PowerISA explicitly says that cfuged/pdepd/pextd is 64-bits only (in the section name), but I can't see the same for vcfuged/vpdepd/vpextd. It's reasonable that no 32-bits CPU implement them, but I don't know about 32-bit code on 64-bit CPU.

**Comment by mferst** [comment]
About split between commits: I'm OK with both options, it's just that upstream seems to prefer smaller commits. I feel that in the addi/addis/paddi and cmp/cmpi/cmpl/cmpli case they where more closely related. E.g. cntlzdm/cnttzdm don't share code, and code shared by pdepd/pextd is also used in unrelated instructions, like cfuged.

File: target/ppc/translate/vector-impl.c.inc
Comment (billionai) [reply target/ppc/translate/vector-impl.c.inc:59]: To add to this discussion, I'm also not sure about REQUIRE_64BIT. Stuff like plfd, for example, isn't obvious that needs to be running in 64 bit mode to be correct. Could it be enough to add a requirement to the ISA 3.1?

**Comment by billionai** [comment]
Are we waiting for something to merge this? If we're waiting for more reviews, I can say that it LGTM

**Comment by mferst** [comment]
Merged in upstream.

**Pull request closed without merging**
